calico/felix/plugins/fiptgenerator.py

```diff
@@ -0,0 +1,731 @@
+            "outbound",
+            expected_mac=mac,
+        )
+
+        updates = {to_chain_name: to_chain, from_chain_name: from_chain}
+        deps = {to_chain_name: to_deps, from_chain_name: from_deps}
+        return updates, deps
+
+    def profile_chain_names(self, profile_id):
+        """
+        Returns the set of chains belonging to a given profile.  This is used
+        e.g. to identify the set of chains that should be deleted to clean up
+        a profile.
+
+        :param profile_id: The profile ID we want to know the chains for.
+        :returns set[string]: the set of chain names
+        """
+        return set([self._profile_to_chain_name("inbound", profile_id),
+                    self._profile_to_chain_name("outbound", profile_id)])
+
+    def profile_updates(self, profile_id, profile, ip_version, tag_to_ipset,
+                        on_allow="ACCEPT", on_deny="DROP", comment_tag=None):
+        """
+        Generate a set of iptables updates that will program all of the chains
+        needed for a given profile.
+
+        :returns Tuple: updates, deps
+        """
+
+        # Generates an inbound and an outbound chain for each profile.
+        # Within each chain, the logic is as follows:
+        # * Start by marking all packets.
+        # * If we hit an allow rule, we'll return with the mark still set, to
+        #   indicate that we matched.
+        # * If we hit a deny rule, we'll drop the packet immediately.
+        # * If we reach the end of the chain, we'll un-mark the packet again to
+        #   indicate no match.
+
+        updates = {}
+        deps = {}
+
+        for direction in ("inbound", "outbound"):
+
+            chain_name = self._profile_to_chain_name(direction, profile_id)
+            rules_key = "%s_rules" % direction
+            rules = profile.get(rules_key, [])
+
+            fragments = [
+                '--append %s --jump MARK --set-mark 1' % chain_name
+            ]
+            for r in rules:
+                rule_version = r.get('ip_version')
+                if rule_version is None or rule_version == ip_version:
+                    fragments.extend(self._rule_to_iptables_fragments(
+                        chain_name,
+                        r,
+                        ip_version,
+                        tag_to_ipset,
+                        on_allow=on_allow,
+                        on_deny=on_deny))
+
+            # If we get to the end of the chain without a match, we remove the
+            # mark again to indicate that the packet wasn't matched.
+            fragments.append(
+                '--append %s '
+                '--match comment --comment '
+                '"No match, fall through to next profile" '
+                '--jump MARK --set-mark 0' % chain_name
+            )
+            updates[chain_name] = fragments
+
+        return updates, deps
+
+    def drop_rules(self, ip_version, chain_name, rule_spec, comment,
```

Comment:
Please change rule_spec and comment to be keyword arguments with default=`None`.  It's bad practice in python to have None-able positionable arguments since, when they're set to None, the only way to tell what they are is to look up the definition and count.